hs2-types/src/Hs2/Types/Type.hs

```diff
@@ -0,0 +1,24 @@
+module Hs2.Types.Type where
+
+type Cxt = Type
+
+data PromotableType
+  = ConTy Name
+  | VarTy Name
+  | TupleTy Int
+
+data Type
+  = ForallTy [Name] Type
+  | ConstrainTy Cxt Type
+  | AppTy Type Type
+  | RegularTy PromotableType
+  | PromotedTy PromotableType
```

Comment:
Maybe we should add row types while we're here. @jvanbruegge, what should an ideal row type look like?